File: Week_01/cpp/src/_234_PalindromeLinkedList.cpp
#include <iostream>

struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};
class Solution
{
public:
    bool isPalindrome(ListNode *head)
    {
        if (not head or not head->next)
            return true;
        // 找到中间节点
        auto p = head, q = head;
        while (p->next && p->next->next)
        {
            p = p->next->next;
            q = q->next;
        }
        // 反转后半段链表
        ListNode *h = NULL;
        p = q->next;
        while (p)
        {
            auto t = p->next;
            p->next = h;
            h = p;
            p = t;
        }
        // 比较两个链表是否相同
        p = head;
        while (h)
        {
            if (h->val != p->val)
            {
                return false;
            }
            h = h->next;
            p = p->next;
        }
        return true;
    }
};

File: Week_01/kotlin/src/_234_PalindromeLinkedList.kt
class _234_PalindromeLinkedList_ {
    /**
     * Example:
     * var li = ListNode(5)
     * var v = li.`val`
     * Definition for singly-linked list.
     * class ListNode(var `val`: Int) {
     *     var next: ListNode? = null
     * }
     */
    class Solution {
        fun isPalindrome(head: ListNode?): Boolean {
            if (head?.next == null) return true
            // 找到中间节点
            var p = head
            var q = head
            while (p?.next != null && p?.next?.next != null) {
                p = p?.next?.next
                q = q?.next
            }
            // 反转链表
            p = null
            q = q?.next
            while (q != null) {
                val t = q?.next
                q?.next = p
                p = q
                q = t
            }
            // 比较链表
            q = head
            while (p != null) {
                if (q?.`val` != p?.`val`) return false
                q = q?.next
                p = p?.next
            }
            return true
        }
    }
}